Userland/Libraries/LibSoftGPU/PixelConverter.cpp

```diff
@@ -0,0 +1,440 @@
+
+    // "By default the components are laid out from msb (most-significant bit) to lsb (least-significant bit)"
+    FloatVector4 components;
+    for (auto i = 0; i < 4; ++i) {
+        auto bitwidth = bitfields[i];
+        if (bitwidth == 0)
+            break;
+        remaining_width -= bitwidth;
+        components[i] = unsigned_to_float((value >> remaining_width) & ((1 << bitwidth) - 1), bitwidth);
+    }
+    return components;
+}
+
+template<>
+constexpr FloatVector4 extract_component_values(Span<float> data_values, GPU::PixelType const&)
+{
+    FloatVector4 components;
+    for (size_t i = 0; i < data_values.size(); ++i)
+        components[i] = data_values[i];
+    return components;
+}
+
+template<typename T>
+static FloatVector4 pixel_values_to_components(Span<T> values, GPU::PixelType const& pixel_type)
+{
+    // Deconstruct read value(s) into separate components
+    auto components = extract_component_values(values, pixel_type);
+    if (pixel_type.components_order == GPU::ComponentsOrder::Reversed)
+        components = { components[3], components[2], components[1], components[0] };
+
+    // Reconstruct component values in order
+    auto component_values = decode_component_order_for_format(components, pixel_type.format);
+    component_values.clamp(0.f, 1.f);
+    return component_values;
+}
+
+FloatVector4 PixelConverter::read_pixel(u8 const** input_data)
+{
+    auto read_components = [&]<typename S, typename O>() {
+        Array<O, 4> values;
+        auto number_of_values = read_pixel_values<S, O>(*input_data, values, m_input_specification);
+        *input_data += number_of_values * sizeof(O);
+        return pixel_values_to_components(values.span().trim(number_of_values), m_input_specification.pixel_type);
+    };
+    switch (m_input_specification.pixel_type.data_type) {
+    case GPU::PixelDataType::Byte:
+        return read_components.template operator()<i8, i8>();
+    case GPU::PixelDataType::Float:
+        return read_components.template operator()<float, float>();
+    case GPU::PixelDataType::HalfFloat:
+        return read_components.template operator()<u16, float>();
+    case GPU::PixelDataType::Int:
+        return read_components.template operator()<i32, i32>();
+    case GPU::PixelDataType::Short:
+        return read_components.template operator()<i16, i16>();
+    case GPU::PixelDataType::UnsignedByte:
+        return read_components.template operator()<u8, u8>();
+    case GPU::PixelDataType::UnsignedInt:
+        return read_components.template operator()<u32, u32>();
+    case GPU::PixelDataType::UnsignedShort:
+        return read_components.template operator()<u16, u16>();
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+static constexpr void write_pixel_data(u8** output_data, float value, GPU::ImageDataLayout layout)
+{
+    auto write_value = [&output_data, &layout]<typename T>(T value) -> void {
+        if (layout.pixel_type.bits == GPU::PixelComponentBits::AllBits) {
```

Comment:
Note to self: remove this and the `layout` argument; `write_pixel_data` is only invoked for `::AllBits`